### PR TITLE
Add new data source type aws_autoscaling_group

### LINKED ISSUE #6758: Feature request: Data Source for : autoscaling_group
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Feature request to add autoscaling_group as a data source. There is this (https://www.terraform.io/docs/providers/aws/d/autoscaling_groups.html) but that does not export information I'm after, e.g. min_size.

Data source could take `arn` as input. API already provides this: https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_DescribeAutoScalingGroups.html

It should support all of the attributes currently provided when creating an autoscaling_group:
```
id - The autoscaling group id.
arn - The ARN for this AutoScaling Group
availability_zones - The availability zones of the autoscale group.
min_size - The minimum size of the autoscale group
max_size - The maximum size of the autoscale group
default_cooldown - Time between a scaling activity and the succeeding scaling activity.
name - The name of the autoscale group
health_check_grace_period - Time after instance comes into service before checking health.
health_check_type - "EC2" or "ELB". Controls how health checking is done.
desired_capacity -The number of Amazon EC2 instances that should be running in the group.
launch_configuration - The launch configuration of the autoscale group
vpc_zone_identifier (Optional) - The VPC zone identifier
load_balancers (Optional) The load balancer names associated with the autoscaling group.
target_group_arns (Optional) list of Target Group ARNs that apply to this AutoScaling Group
```

### New or Affected Resource(s)

* aws_XXXXX

### Potential Terraform Configuration

```hcl
data "aws_autoscaling_group" "foobar" {
  arn = "...."
}

```

### References
This is _not_ to be confused with aws_autoscaling_group**s**: https://www.terraform.io/docs/providers/aws/d/autoscaling_groups.html

### PR DESCRIPTION
**Pull request by deberon**
Fixes #6758 

Changes proposed in this pull request:

* Add a new data source type called aws_autoscaling_group
* Add documentation for new data source type

Output from acceptance testing:

```
$ TESTARGS="-run TestAccAwsAutoScalingGroupDataSource_basic" make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run TestAccAwsAutoScalingGroupDataSource_basic -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAwsAutoScalingGroupDataSource_basic
=== PAUSE TestAccAwsAutoScalingGroupDataSource_basic
=== CONT  TestAccAwsAutoScalingGroupDataSource_basic
--- PASS: TestAccAwsAutoScalingGroupDataSource_basic (49.77s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       49.786s
```

### CODE REVIEW

**Comment by deberon** [comment]
And for the sake of consistency, here is the output of the acceptance testing after my second commit:
```
$ TESTARGS="-run TestAccAwsAutoScalingGroupDataSource_basic" make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run TestAccAwsAutoScalingGroupDataSource_basic -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAwsAutoScalingGroupDataSource_basic
=== PAUSE TestAccAwsAutoScalingGroupDataSource_basic
=== CONT  TestAccAwsAutoScalingGroupDataSource_basic
--- PASS: TestAccAwsAutoScalingGroupDataSource_basic (49.55s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       49.565s
```

File: aws/data_source_aws_autoscaling_group_test.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  owners = ["099720109477"] # Canonical
+}
+
+data "aws_availability_zones" "available" {
+  state = "available"
+}*/
+
+resource "aws_launch_configuration" "data_source_aws_autoscaling_group_test" {
+  name          = "data_source_aws_autoscaling_group_test"
+  image_id      = "ami-090cf2033b0eefe38"
```
Comment (bflad) [inline aws/data_source_aws_autoscaling_group_test.go]: We should either remove the commented `aws_ami` data source configuration above or preferably uncomment it and use it:

```hcl
  image_id      = "${data.aws_ami.ubuntu.id}"
```

File: aws/data_source_aws_autoscaling_group_test.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+resource "aws_autoscaling_group" "foo" {
+  name                      = "asg_test_foo"
+  max_size                  = 0
+  min_size                  = 0
+  health_check_grace_period = 300
+  health_check_type         = "ELB"
+  desired_capacity          = 0
+  force_delete              = true
+  launch_configuration      = "${aws_launch_configuration.data_source_aws_autoscaling_group_test.name}"
+  availability_zones	    = ["us-west-2a"]
```
Comment (bflad) [inline aws/data_source_aws_autoscaling_group_test.go]: We should either remove the commented `aws_availability_zones` data source configuration above or preferably uncomment it and use it:

```hcl
  availability_zones	    = ["${data.aws_availability_zones.available.name[0]}"]
```

Same with the other configuration `aws_autoscaling_group` configuration below.

File: aws/data_source_aws_autoscaling_group_test.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+}*/
+
+resource "aws_launch_configuration" "data_source_aws_autoscaling_group_test" {
+  name          = "data_source_aws_autoscaling_group_test"
+  image_id      = "ami-090cf2033b0eefe38"
+  instance_type = "t2.micro"
+}
+
+resource "aws_autoscaling_group" "foo" {
+  name                      = "asg_test_foo"
```
Comment (bflad) [inline aws/data_source_aws_autoscaling_group_test.go]: The `fmt.Sprintf()` naming randomization should be extended here as well, `name = "%s-foo"` here and `name = "%s-bar"` below.

File: website/docs/d/autoscaling_group.html.markdown [hunk 4]
```diff
@@ -0,0 +1,47 @@
+---
+layout: "aws"
+page_title: "AWS: aws_autoscaling_group"
+sidebar_current: "docs-aws-datasource-autoscaling-group"
```
Comment (bflad) [inline website/docs/d/autoscaling_group.html.markdown]: This documentation page is missing a sidebar link in `website/aws.erb` in the data sources section.

File: website/docs/d/autoscaling_group.html.markdown [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+---
+
+# Data Source: aws_autoscaling_group
+
+Use this data source to get information on an existing autoscaling group.
+
+## Example Usage
+
+```hcl
+data "autoscaling_group" "foo" {
```
Comment (bflad) [inline website/docs/d/autoscaling_group.html.markdown]: 😎 Typo

```suggestion
data "aws_autoscaling_group" "foo" {
```

File: aws/data_source_aws_autoscaling_group.go [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+	if err := setLoadBalancers(d, group); err != nil {
+		return err
+	}
+	d.Set("max_size", group.MaxSize)
+	d.Set("min_size", group.MinSize)
+	d.Set("new_instances_protected_from_scale_in", group.NewInstancesProtectedFromScaleIn)
+	d.Set("placement_group", group.PlacementGroup)
+	d.Set("service_linked_role_arn", group.ServiceLinkedRoleARN)
+	d.Set("status", group.Status)
+	d.Set("target_group_arns", group.TargetGroupARNs)
```
Comment (bflad) [inline aws/data_source_aws_autoscaling_group.go]: [`group.TargetGroupARNs` is a `[]*string`](https://docs.aws.amazon.com/sdk-for-go/api/service/autoscaling/#Group) and needs to be handled with `aws.StringValueSlice()` (or will silently fail to set anything in the Terraform state):

```suggestion
	if err := d.Set("target_group_arns", aws.StringValueSlice(group.TargetGroupARNs)); err != nil {
		return fmt.Errorf("error setting target_group_arns: %s", err)
	}
```

File: aws/data_source_aws_autoscaling_group.go [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+	d.Set("termination_policies", group.TerminationPolicies)
```
Comment (bflad) [inline aws/data_source_aws_autoscaling_group.go]: [`group.TerminationPolicies` is a `[]*string`](https://docs.aws.amazon.com/sdk-for-go/api/service/autoscaling/#Group) and needs to be handled with `aws.StringValueSlice()` (or will silently fail to set anything in the Terraform state):

```suggestion
	if err := d.Set("termination_policies", aws.StringValueSlice(group.TerminationPolicies)); err != nil {
		return fmt.Errorf("error setting termination_policies: %s", err)
	}
```

File: aws/data_source_aws_autoscaling_group_test.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
+package aws
+
+import (
+	"testing"
+
+	//"fmt"
```
Comment (bflad) [inline aws/data_source_aws_autoscaling_group_test.go]: Commented imports should be removed. Tools like [`goimports`](https://godoc.org/golang.org/x/tools/cmd/goimports) generally have editor plugins which can automatically handle Go imports additions/removals/ordering during save. 😄

**Comment by bflad** [comment]
> It was my first time writing in Go, could you tell?

Hah, I don't judge, but you did pretty good considering that new fact to me. Setting up a Go environment for the first time, compiling a large project like this one, and developing/running integration testing is no small feat. Great job!

> How do you prefer me to submit my patches? Should I squash the newer commits or do you guys like the verbose history?

Verbose history or squashing is fine. We'll merge the commits either which way.

File: aws/data_source_aws_autoscaling_group.go [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+
+	if len(result.AutoScalingGroups) < 1 {
+		return fmt.Errorf("Your query did not return any results. Please try a different search criteria.")
+	}
+
+	if len(result.AutoScalingGroups) > 1 {
+		return fmt.Errorf("Your query returned more than one result. Please try a more " +
+			"specific search criteria.")
+	} else {
+		group = result.AutoScalingGroups[0]
```
[the remark this answers is not in the document]
Comment (deberon) [reply aws/data_source_aws_autoscaling_group.go]: Ah I had no idea about this functionality in Go. Thanks!

**Comment by deberon** [comment]
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run TestAccAwsAutoScalingGroupDataSource_basic -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAwsAutoScalingGroupDataSource_basic
=== PAUSE TestAccAwsAutoScalingGroupDataSource_basic
=== CONT  TestAccAwsAutoScalingGroupDataSource_basic
--- PASS: TestAccAwsAutoScalingGroupDataSource_basic (51.30s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       (cached)

Also I see there's a conflict on website/aws.erb. I tried to lay my changes over the most current master copy but it's a popular repo.

**Comment by bflad** [comment]
This has been released in [version 1.54.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1540-december-21-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**